### PR TITLE
Template structure simplification

### PR DESCRIPTION
**Pull request by nelimee**
It has been a few weeks I am thinking about that, and a discussion with Kabir a few days ago finally convinced me that refactoring the Template hierarchy was a good idea.

The main reason for that refactor is the unneeded complexity of the old `Template` code: we only need to represent "simple" templates such as logical qubits and their sides. Being able to represent arbitrary templates using the `ComposedTemplate` class was nice, but not a requirement. With the need to query templates about their middle lines, sides, and likely other properties in the future that were not easily implementable (or even correctly defined) for all the possible `ComposedTemplate` instance, removing that intermediary seemed like a necessity.

This (rather large, sorry) commit aims at greatly simplifying the `Template`-related module. The interfaces are left unchanged, which reduces by a lot the possibility of introducing errors with this refactor and reduces the number of changes.

Note that this is only a draft pull request, I **really** want other maintainer opinions on that before even considering it mergeable. 

Note also that both #71 and #274 will be impacted by this refactor.

Finally, this is also an experiment to test another idea for automatic detector computation. I will use that PR as a base for another PR on that, and see if it turns out to be a nice approach.

### CODE REVIEW

**Review by inmzhang — COMMENTED** [review]
The removal of `ComposedTemplate` indeed simplifies things a lot. And currently I haven't seen any bottlenecks of this simplification considering the current workflow. A few questions in my mind:
1. How do you plan to construct more complex templates like the corners/junctions? I assume that for each "composed template" now we need to implement a rather large `instantiate` method to specify all the plaquette indices.
2. I remembered that the complexity of `PiecewiseLinearFunction` mainly comes from the usage in `ComposedTemplate`. If it does not exist anymore, is it possible to simplify the `scale.py`/`interval.py` as well?

**Comment by Gistbatch** [comment]
I'm not sure what kind of complex structures we would need, but I think both corners and junctions can be represented as a block graph?

**Comment by inmzhang** [comment]
Yes they can be represented by block graph. But when it comes to instantiate them into circuits, there is the possibility that we need to change the underlying template for more complex circuits to fit in. For example, corner/junctions in space need special care about the plaquettes arrangement for the circuit to be distance-preserving.

**Comment by nelimee** [comment]
> 1. How do you plan to construct more complex templates like the corners/junctions? I assume that for each "composed template" now we need to implement a rather large `instantiate` method to specify all the plaquette indices.

The problem is the same with the current architecture. The main complexity is due to the center `2k x 2k` template that has a "less regular" pattern in the case of corner or junctions. Implementing a corner / junction with the `ComposedTemplate` API still needs to specify that `2k x 2k` template by hand, which is the main part of complexity. 
In other words, because the border implementation is not really complex (and can still be factored out in a separate function), I do not see any added complexity by this approach, the same things have to be done anyway in both approaches.



> I remembered that the complexity of `PiecewiseLinearFunction` mainly comes from the usage in `ComposedTemplate`. If it does not exist anymore, is it possible to simplify the `scale.py`/`interval.py` as well?

This is very much true, I will try to remove that now.

**Comment by nelimee** [comment]
In order to demonstrate that it can be done, I implemented the 4-way junction as a new template. Note that 2/3-way junctions are not implemented yet, and there are enough questions we need to collectively answer on them to justify another PR just for that.
I think this PR is ready, but before merging I would like @KabirDubey to react here as I know he was working on #71, which is greatly impacted by this PR.

**Comment by nelimee** [comment]
Now that I included more tests I think this PR is ready to be reviewed. As soon as it is accepted, I will merge this one and just after merge #318. 
Then, I will need to fix any potential merge conflict with #305 and merge #305.
The next steps will be:
1. Finish #319 and merge it,
2. Start and finish a translation method from `AbstractObservable` instances to actual `Observable`, or even better to a new `ScalingObservable`.

**Review by inmzhang — APPROVED** [review]
I'll approve this without carefully checking the 4-way junction implementation. But I regard it is more of a showcase of the `Template` and the correctness does not matter so much for now. And it just looks correct...

**Comment by nelimee** [comment]
This is also how I see it: a showcase that will need to be checked and improved. Thanks for the quick approval !

File: src/tqec/templates/display.py
Comment (KabirDubey) [inline src/tqec/templates/display.py:1]: Do we need new display functions for the refactored `Template`? I can work on this (make a new issue and close issue #71). I'm still in the process of understanding the refactoring.

Comment (nelimee) [reply src/tqec/templates/display.py:1]: > Do we need new display functions for the refactored `Template`?

I am not sure. I think it depends nearly entirely on whether or not we plan to eventually have a frontend that will need to visually represent templates. Another point is that it would be nice to have nice looking images for the documentation. 

So let's say that we might need it, but it is clearly not sure and not a priority. If you still want to work on it, please feel free to do so :)

**Pull request merged**
